pkg/proxy/envoyproxy.go

```diff
@@ -0,0 +1,142 @@
+// Copyright 2016-2017 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package proxy
+
+import (
+	"fmt"
+	"net/http"
+	"net/url"
+	"sync"
+	"time"
+
+	"github.com/cilium/cilium/pkg/envoy"
+	"github.com/cilium/cilium/pkg/policy"
+	"github.com/cilium/cilium/pkg/proxy/accesslog"
+
+	log "github.com/sirupsen/logrus"
+	"github.com/spf13/viper"
+)
+
+// the global Envoy instance
+var envoyProxy *envoy.Envoy
+
+// EnvoyRedirect implements the Redirect interface for a l7 proxy
+type EnvoyRedirect struct {
+	id      string
+	toPort  uint16
+	ingress bool
+	source  ProxySource
+}
+
+// ToPort returns the redirect port of an OxyRedirect
+func (r *EnvoyRedirect) ToPort() uint16 {
+	return r.toPort
+}
+
+// IsIngress returns true if the redirect is for ingress, and false for egress.
+func (r *EnvoyRedirect) IsIngress() bool {
+	return r.ingress
+}
+
+func (r *EnvoyRedirect) getSource() ProxySource {
+	return r.source
+}
+
+var envoyOnce sync.Once
+
+// createOxyRedirect creates a redirect with corresponding proxy
+// configuration. This will launch a proxy instance.
+func createEnvoyRedirect(l4 *policy.L4Filter, id string, source ProxySource, to uint16) (Redirect, error) {
+	envoyOnce.Do(func() {
+		// Start Envoy on first invocation
+		envoyProxy = envoy.StartEnvoy(true, 0, viper.GetString("state-dir"), viper.GetString("state-dir"))
+	})
+
+	if envoyProxy != nil {
```

Comment:
Can this race with the envoyOnce above? I can't think of a way to sequence this without a lock, unfortunately :/

Reply:
According to sync spec "no call to Do returns until the one call to f returns", so the locking here is implicit.

Reply:
ooh, I didn't know that! really handy to know.